### PR TITLE
Update HandlerContext signal

### PR DESCRIPTION
**Pull request by johynpapin**
This PR renames the AbortSignal property in the HandlerContext from `deadline` to a more generic `signal`. 

It also adds a signal to universal server requests and links it to the handler context signal. 

We expect server plugins to trigger the AbortSignal of the request when a request is finished or aborted prematurely. 

As a result, service implementations can now rely on the signal of the HandlerContext to clean up, whether the RPC timed out (via a deadline), whether the client canceled the request, or whether the request finished normally.

### CODE REVIEW

**Comment by timostamm** [comment]
Thank you for the PR, Johyn! We'll take a look soon.

**Comment by johynpapin** [comment]
After reading #591, I wonder if we could just abort the `deadline` signal when the request is closed instead of adding another signal.

File: packages/connect-node/src/node-universal-handler.ts
Comment (timostamm) [inline packages/connect-node/src/node-universal-handler.ts:93]: We have to honor HTTP/2 error codes here. We can use `connectErrorFromH2ResetCode()` added in https://github.com/bufbuild/connect-es/pull/619.

For example:

```ts
  if ("stream" in nodeRequest) {
    // HTTP/2 has error codes we want to honor
    nodeRequest.on("close", () => {
      const err = connectErrorFromH2ResetCode(nodeRequest.stream.rstCode);
      if (err) {
        ac.abort(err);
      } else {
        ac.abort();
      }
    });
  } else {
    // HTTP/1.1 does not
    nodeRequest.on("close", () => ac.abort());
  }
```

Yes, I expect the "close" event to always fire.

Aborting with a reason is not supported in Node.js v16, but it will be EOL soon. For the interim, it will be fine if v16 does not bubble up the code correctly.

We should update `universalServerRequestFromFetch()` as well, and use the signal from the fetch Request.

File: packages/connect/src/implementation.ts
Comment (timostamm) [inline packages/connect/src/implementation.ts:103]: I am afraid I made this a bit more difficult with https://github.com/bufbuild/connect-es/pull/591 🙈 

Yes, it would be nice if we just had a single AbortSignal, but you need an Abort_Controller_ to trigger an AbortSignal, and I think `universalRequestFromNodeRequest()` should neither accept nor return an AbortController. That means we cannot avoid linking AbortSignals.

But it's not too bad: https://github.com/bufbuild/connect-es/pull/623 added `createLinkedAbortController()` for this purpose. We can merge the signal from the deadline / shutdown with the one from the request before passing it to the HandlerContext.

**Review by timostamm — COMMENTED** [review]
Hey @johynpapin, took me a minute to get back to this :blush:

Adding the signal to the UniversalServerRequest, aborting it in node-universal-handler.ts, and making it available in the HandlerContext is a good call. 

There is one small but important bit though: HTTP/2 has a set of error codes ([spec](https://httpwg.org/specs/rfc7540.html#ErrorCodes)), and gRPC defines a mapping between these codes, and gRPC error codes ([spec](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#errors)). If a HTTP/2 client cancels a request, it should close the stream with the HTTP/2 code CANCEL (implemented in https://github.com/bufbuild/connect-es/pull/552). The server should pick this code up, and convert it to a Connect/gRPC code "canceled".

If you can spare the time to update the PR, we'd be super happy to have it 🙂

**Comment by johynpapin** [comment]
Hey @timostamm! :wave:

Thank you very much for this detailed answer. I suspected that there was a problem with the error handling, it's clearer to me now.

I will take the time to understand and make these changes! :smile:

File: packages/connect/src/implementation.ts
Comment (johynpapin) [reply packages/connect/src/implementation.ts:103]: If I understand correctly it is necessary to wait until #623 is merged to implement this change? :open_mouth:

Comment (timostamm) [reply packages/connect/src/implementation.ts:107]: Yes, https://github.com/bufbuild/connect-es/pull/623 should be merged soon 🙂

File: packages/connect/src/protocol/universal-fetch.ts
Comment (timostamm) [reply packages/connect/src/protocol/universal-fetch.ts:94]: I believe it is the best we can do. For context, Deno has been starting to make use of the property on servers in https://github.com/denoland/deno/pull/17822.

File: packages/connect/src/protocol/universal-handler-client.ts
Comment (timostamm) [reply packages/connect/src/protocol/universal-handler-client.ts:47]: Yes, I believe this should work. We will add more test coverage for this function soon, and can fix up if necessary.

File: packages/connect/src/protocol/universal.ts
Comment (timostamm) [reply packages/connect/src/protocol/universal.ts:67]: I think it's fair to use just `signal`, like the fetch API Request 👍

File: packages/connect/src/implementation.ts
Comment (timostamm) [inline packages/connect/src/implementation.ts:115]: Good idea to link the signals here 👍 
Can you simply put `createLinkedAbortController(deadline, signal)` on this line?

File: packages/connect/src/implementation.ts
Comment (timostamm) [inline packages/connect/src/implementation.ts:77]: We wrap JSDoc at 80 columns:

```suggestion
  /**
   * An AbortSignal that is aborted when the connection with the client is closed 
   * or when the deadline is reached.
   */
```

**Review by timostamm — COMMENTED** [review]
Looks great! 

Do not worry about the CI failure. The fork does not have access to the secret that is required for browserstack. As long as `make` passes (as it did), everything is fine, since we are not touching code that directly affects @bufbuild/connect-web.

I'm going to think about test coverage tomorrow morning and get back to you.

File: packages/connect/src/implementation.ts
Comment (johynpapin) [reply packages/connect/src/implementation.ts]: Yes!

**Review by timostamm — APPROVED** [review]
Thank for for the contribution, Johyn, LGTM! 

I'm going to update the PR description, and also update the description of https://github.com/bufbuild/connect-es/pull/591 for the renamed HandlerContext property.

**Review by timostamm — APPROVED** [review]
Thank for for the contribution, Johyn, LGTM! 

I'm going to update the PR description, and also update the description of https://github.com/bufbuild/connect-es/pull/591 for the renamed HandlerContext property.

**Pull request merged**
